In [1]:
import pandas as pd

# Reading and displaying the dataset

In [2]:
data2 = pd.read_csv('poke.csv')
data2

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

# Installing and Importing altair package for visualisation

In [3]:
#pip install altair
import altair as alt

# Displaying the columns(Features) in the dataset

In [4]:
data2.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

# Checking for null values in the dataset

In [5]:
data2.isnull()

#   Name  Type 1  Type 2  Total     HP  Attack  Defense  Sp. Atk  \
0    False  False   False   False  False  False   False    False    False   
1    False  False   False   False  False  False   False    False    False   
2    False  False   False   False  False  False   False    False    False   
3    False  False   False   False  False  False   False    False    False   
4    False  False   False    True  False  False   False    False    False   
..     ...    ...     ...     ...    ...    ...     ...      ...      ...   
795  False  False   False   False  False  False   False    False    False   
796  False  False   False   False  False  False   False    False    False   
797  False  False   False   False  False  False   False    False    False   
798  False  False   False   False  False  False   False    False    False   
799  False  False   False   False  False  False   False    False    False   

     Sp. Def  Speed  Generation  Legendary  
0      False  False       False      False  
1      False  False       False      False  
2      False  False       False      False  
3      False  False       False      False  
4      False  False       False      False  
..       ...    ...         ...        ...  
795    False  False       False      False  
796    False  False       False      False  
797    False  False       False      False  
798    False  False       False      False  
799    False  False       False      False  

[800 rows x 13 columns]

## To get a better idea of the columns containing null values

In [6]:
info = data2.info()
info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


Here we can see that column 3 - "Type 2" has null values. 

In [7]:
data2['Type 2'].isna().sum()

386

<b> Column ['Type 2'] has 386 null/missing values which is ~50% of the rows in the dataset</b>

# Getting an idea of the distribution and statistics of the quantitative variables in the dataset

In [8]:
description = data2.describe()
description

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

'Generation' is actually a categorical variable that classifies the different pokemon into generations but has numerical values. to veriy this, we check the unique values of 'Generation'.

In [9]:
data2['Generation'].unique()

array([1, 2, 3, 4, 5, 6])

We see that 'Generation' has 6 unique values

# Displaying the first 10 rows of the dataset

In [10]:
data2.head(10)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9  7                   Squirtle  Water     NaN    314  44      48       65   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False  
6      109       85    100           1      False  
7      130       85    100           1      False  
8      159      115    100           1      False  
9       50       64     43           1      False

# Tasks to visualise
<br>

### We have decided to visualise the following tasks in this pipeline

<br>
1. Number of Pokemon in each Generation
<br>
2. Type distribution with each generation 
<br>
3. Attack Distribution by Type 1 and Legendary
<br>
4. Defense distribution by Type 1 and Legendary 
<br>
5. Top 10 non legendary pokemon with highest stats
<br>
6. Legendary Pokemon stats by Generation
<br>



# Task 1 -  Number of Pokemon in each Generation 

In [11]:
#Representing count(Pokemon) with respect to generation

alt.Chart(data2).mark_bar().encode(x = 'Generation:N',
                                           y = 'count_name:Q',
                                           color = 'Generation:N').transform_aggregate(count_name = 'count(Name):Q', 
                                        groupby=["Generation"]).properties(
                                                                title = 'Pokemon by Generation', 
                                                                width = 500, height = 500).configure_title(
                                                                fontSize = 20, font='Times New Roman').configure_legend(labelFontSize = 20, 
                                                                titleFontSize= 20).configure_axis(
                                                                labelFontSize = 15, titleFontSize = 15).interactive()

alt.Chart(...)

# Task 2- Type distribution For Each Generation

In [12]:
brush = alt.selection_single(fields = ['Type 1'])

points = alt.Chart(data2).mark_bar().encode(
                x = alt.X('Generation:N'),
                y = alt.Y('count_Pokemon:Q'),
                color = alt.condition(brush, 'Type 1:N',
                                      alt.value('gray')),
                tooltip = ['Generation', 'Type 1', 'Name']
).add_selection(brush).transform_aggregate(
    count_Pokemon = 'count(Name):Q', groupby=['Generation', 'Type 1', 'Name']
).properties(width = 500, height = 500, title = 'Element Type 1 of Pokemon by Generation')


bars = alt.Chart(data2).mark_bar().encode(
y = 'count(Type 1):O',
x = 'Type 1:N',
color = alt.condition(brush,'Type 1:N',alt.value('gray')),
tooltip = ['count(Type 1)', 'Type 1']
).properties(title = 'Count of Selected Pokemon').transform_filter(brush).add_selection(brush)

points | bars

alt.HConcatChart(...)

# Task 3 - Attack distribution by Type 1 and Legendary

In [13]:
#min-max boxplot
alt.Chart(data2).mark_boxplot(extent='min-max').encode(
    x='Type 1:N',
    y ='Attack:Q',
    color = "Legendary:N").configure_legend(labelFontSize = 20, titleFontSize= 20).configure_axis(
                        labelFontSize = 15, titleFontSize = 15).properties(width = 500, height = 500,
                                     title = 'Attack Range of Pokemon by Type 1').configure_title(
                                     fontSize = 20, font = 'Times New Roman').interactive()

alt.Chart(...)

# Task 4 - Defense Distribution by Type 1 and Legendary

In [14]:
#min-max boxplot
alt.Chart(data2).mark_boxplot(extent='min-max').encode(
    x='Type 1:N',
    y ='Defense:Q',
    color = "Legendary:N").properties(width = 500, height = 500,
                            title = "Defense range of Pokemon by type 1").configure_title(
                            fontSize = 20, font = 'Times New Roman').configure_legend(
                            labelFontSize = 20, titleFontSize= 20).configure_axis(
                            labelFontSize = 15, titleFontSize = 15).interactive()

alt.Chart(...)

# Task 5 - Top 10 non legendary pokemon with highest stats

In [15]:
#Sorting by stats
data_pokemon = data2[data2['Legendary']== False]

#Extracting a subset of data and sorting
data_pokemon = data_pokemon.sort_values(by = ['Total'],ascending = False)
data_pokemon = data_pokemon[~data_pokemon['Name'].str.contains("Mega")].head(10)

#Renaming column names of Sp. Atk and Sp. Def
data_pokemon.rename(columns={'Sp. Atk':'Special_Attack', 'Sp. Def':'Special_Defense'}, inplace=True)
data_pokemon.head()

#                     Name   Type 1    Type 2  Total   HP  Attack  \
313  289                  Slaking   Normal       NaN    670  150     160   
716  648  MeloettaPirouette Forme   Normal  Fighting    600  100     128   
548  490                  Manaphy    Water       NaN    600  100     100   
546  488                Cresselia  Psychic       NaN    600  120      70   
696  635                Hydreigon     Dark    Dragon    600   92     105   

     Defense  Special_Attack  Special_Defense  Speed  Generation  Legendary  
313      100              95               65    100           3      False  
716       90              77               77    128           5      False  
548      100             100              100    100           4      False  
546      120              75              130     85           4      False  
696       90             125               90     98           5      False

In [16]:
#Heatmap to show the stats
alt.Chart(data_pokemon).mark_rect().encode(
        x = alt.X('Name:N', sort = '-y'),
        y = alt.Y('Total:Q'),
        tooltip=['Name','Attack','Defense','Special_Attack','Special_Defense', 'Type 1','Type 2'],
        color=alt.Color('Generation:N')
).properties(width = 400, 
             height = 400, title = 'Stats of top 10 Non Legendary Pokemon').configure_title(
             fontSize = 20, font = 'Times New Roman').configure_legend(labelFontSize = 20, 
                                                                       titleFontSize= 20).configure_axis(
                                                                       labelFontSize = 15, titleFontSize = 15)

alt.Chart(...)

In [17]:
#Extracting Legendary pokemon Stats from the dataset
data_pokemon_legendary = data2[data2['Legendary']==True]
data_pokemon_legendary = data_pokemon_legendary[~data_pokemon_legendary['Name'].str.contains("Mega")]

#Renaming the columns
data_pokemon_legendary.rename(columns={'Sp. Atk':'Special_Attack', 'Sp. Def':'Special_Defense'}, inplace=True)
data_pokemon_legendary.head()

#      Name    Type 1  Type 2  Total   HP  Attack  Defense  \
156  144  Articuno       Ice  Flying    580   90      85      100   
157  145    Zapdos  Electric  Flying    580   90      90       85   
158  146   Moltres      Fire  Flying    580   90     100       90   
162  150    Mewtwo   Psychic     NaN    680  106     110       90   
262  243    Raikou  Electric     NaN    580   90      85       75   

     Special_Attack  Special_Defense  Speed  Generation  Legendary  
156              95              125     85           1       True  
157             125               90    100           1       True  
158             125               85     90           1       True  
162             154               90    130           1       True  
262             115              100    115           2       True

# Task 6- Legendary Pokemon stats by Generation

In [18]:
alt.Chart(data_pokemon_legendary).mark_rect().encode(
        x=alt.X('Name:N', sort = '-y'),
        y='Total:Q',
        color='Generation:N',
        tooltip=['Name','Attack','Type 1','Type 2','Defense','Special_Attack','Special_Defense', 'Generation'],
).properties(title = 'Stats of Legendary Pokemon by Generation').configure_title(
                        fontSize = 20, font = 'Times New Roman').configure_legend(
                        labelFontSize = 20, titleFontSize= 20).configure_axis(
                        labelFontSize = 15, titleFontSize = 15)

alt.Chart(...)